SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,711 @@
+@property (strong, nonatomic, nullable) SDLTimer *registerTransportTimer;
+
+// A delegate to notify protocol change.
+@property (weak, nonatomic) id<SDLStreamingProtocolDelegate> streamingProtocolDelegate;
+
+// Configuration sent by system; list of transports that are allowed to carry audio service
+@property (strong, nonatomic, nonnull) NSArray<SDLTransportClassBox *> *transportsForAudioService;
+// Configuration sent by system; list of transports that are allowed to carry video service
+@property (strong, nonatomic, nonnull) NSArray<SDLTransportClassBox *> *transportsForVideoService;
+// A map to remember which service is currently running on which transport
+@property (strong, nonatomic) NSMutableDictionary<SDLServiceTypeBox *, SDLTransportClassBox *> *streamingServiceTransportMap;
+
+// IP address of SDL Core (either IPv4 or IPv6 address)
+@property (strong, nonatomic, nullable) NSString *ipAddress;
+// TCP port number of SDL Core. If the information isn't available then TCPPortUnspecified is stored.
+@property (assign, nonatomic) int tcpPort;
+
+@end
+
+@implementation SDLSecondaryTransportManager
+
+#pragma mark - Public
+
+- (instancetype)initWithStreamingProtocolDelegate:(id<SDLStreamingProtocolDelegate>)streamingProtocolDelegate
+                                      serialQueue:(dispatch_queue_t)queue {
+    self = [super init];
+    if (!self) {
+        return nil;
+    }
+
+    _stateMachine = [[SDLStateMachine alloc] initWithTarget:self initialState:SDLSecondaryTransportStateStopped states:[self.class sdl_stateTransitionDictionary]];
+    _stateMachineQueue = queue;
+    _streamingProtocolDelegate = streamingProtocolDelegate;
+
+    _secondaryTransportType = SDLSecondaryTransportTypeDisabled;
+    _transportsForAudioService = @[];
+    _transportsForVideoService = @[];
+    _streamingServiceTransportMap = [@{@(SDLServiceTypeAudio):@(SDLTransportClassInvalid),
+                            @(SDLServiceTypeVideo):@(SDLTransportClassInvalid)} mutableCopy];
+    _tcpPort = TCPPortUnspecified;
+
+    return self;
+}
+
+- (void)startWithPrimaryProtocol:(SDLProtocol *)primaryProtocol {
+    SDLLogD(@"SDLSecondaryTransportManager start");
+
+    // this method must be called in SDLLifecycleManager's state machine queue
+    dispatch_assert_queue(self.stateMachineQueue);
+
+    if (![self.stateMachine isCurrentState:SDLSecondaryTransportStateStopped]) {
+        SDLLogW(@"Secondary transport manager is already started!");
+        return;
+    }
+
+    self.primaryProtocol = primaryProtocol;
+    self.primaryProtocolHandler = [[SDLSecondaryTransportPrimaryProtocolHandler alloc] initWithSecondaryTransportManager:self primaryProtocol:primaryProtocol];
+
+    [self.stateMachine transitionToState:SDLSecondaryTransportStateStarted];
+}
+
+- (void)stop {
+    SDLLogD(@"SDLSecondaryTransportManager stop");
+
+    // this method must be called in SDLLifecycleManager's state machine queue
+    dispatch_assert_queue(self.stateMachineQueue);
```

Comment:
This assertion creates a failure in a unit test `a_lifecycle_manager__when_started__in_the_ready_state__stopping_the_manager__when_receiving_an_unregister_response__should_stop` of SDLLifecycleManagerSpec.

Looking at the code, I think this is rather an issue of `SDLLifecycleManager`. There is a case when `didEnterStateUnregistering` method is called from `SDLResponseDispatcher`, and in that case `didEnterStateUnregistering` method may not be called on the state machine queue. (I think the state machine methods should be always invoked on the dedicated queue.)

Right now I am not sure if this is a real bug or it's just an issue specific to that unit test case.

Reply:
I found another case where an app terminates because of this `dispatch_assert_queue()`. When I disables a transport (like turning off BT), SDLLifecycleManager's `sdl_stopManager:` calls `transitionToState` method on the *main queue*, which eventually triggers SDLSecondaryTransportManager's `startWithPrimaryProtocol` and reaches `dispatch_assert_queue`.
https://github.com/smartdevicelink/sdl_ios/blob/develop/SmartDeviceLink/SDLLifecycleManager.m#L251

I think it is better to make it an error log rather than an assertion, so that developers will not face app terminations during development. I put my change with commit 2eaa2fd. Hopefully I can create another PR to fix the queue issue.

That commit also resolves the unit test failure.

Reply:
The issue of the state machine running on unexpected queues should be addressed by https://github.com/smartdevicelink/sdl_ios/pull/1069.